requests/models.py

```diff
@@ -470,18 +469,12 @@ def prepare_body(self, data, files, json=None):
         self.body = body
 
     def prepare_content_length(self, body):
-        if hasattr(body, 'seek') and hasattr(body, 'tell'):
-            curr_pos = body.tell()
-            body.seek(0, 2)
-            end_pos = body.tell()
-            self.headers['Content-Length'] = builtin_str(max(0, end_pos - curr_pos))
-            body.seek(curr_pos, 0)
-        elif body is not None:
-            l = super_len(body)
-            if l:
-                self.headers['Content-Length'] = builtin_str(l)
-        elif (self.method not in ('GET', 'HEAD')) and (self.headers.get('Content-Length') is None):
-            self.headers['Content-Length'] = '0'
+        if self.headers.get('Content-Length') is None:
+            if body is not None:
+                length = super_len(body)
+                self.headers['Content-Length'] = builtin_str(length)
+            elif self.method not in ('GET', 'HEAD'):
+                self.headers['Content-Length'] = '0'
```

Comment:
Allowing users to override `Content-Length` is a dangerous game. In particular, users who are *wrong* about the header will encounter problems whereby either the server returns 400 Client Error, terminates the connection, or worse, the connection hangs while the server waits for more body. Your example is a great example of this case: the user should be *utterly forbidden* from emitting that request. I have exactly no interest in allowing users to easily emit that request.

The case I *do* care about is the possibility that the user may want to set Content-Length in one specific case: the case where we'd normally set `Transfer-Encoding: chunked`. In this instance we could take this as a signal to avoid chunked transfer encoding. However, in general, we haven't done this consistently and this PR is not the right place to address it.